granulate_utils/metadata/databricks_client.py

```diff
@@ -34,55 +36,99 @@ def __init__(self, logger: logging.LoggerAdapter) -> None:
+                if cluster_metadata := self._cluster_all_tags_metadata():
+                    name = self._get_name_from_metadata(cluster_metadata)
+                    if name:
+                        self.logger.debug("Found name in metadata", job_name=name, cluster_metadata=cluster_metadata)
+                        return name
+                    else:
+                        self.logger.debug("Failed to extract name from metadata", cluster_metadata=cluster_metadata)
+                        return None
+                else:
+                    # No job name yet, retry.
+                    time.sleep(RETRY_INTERVAL_S)
+            except DatabricksJobNameDiscoverException:
+                self.logger.exception("Failed to get Databricks job name")
+                return None
             except Exception:
-                self.logger.exception("Got Exception while collecting Databricks job name.")
+                self.logger.exception("Generic exception was raise during spark job name discovery")
+                return None
+        self.logger.info("Databricks get job name timeout, continuing...")
         return None
 
-    def _get_job_name_impl(self) -> Optional[str]:
-        # Make sure we're running on a databricks machine
-        if not os.path.isfile(DATABRICKS_DEPLOY_CONF_PATH):
+    @staticmethod
+    def _get_name_from_metadata(metadata: Dict[str, str]) -> Optional[str]:
+        if JOB_NAME_KEY in metadata:
+            return str(metadata[JOB_NAME_KEY]).replace(" ", "-").lower()
+        return None
+
+    def _cluster_all_tags_metadata(self) -> Optional[Dict[str, str]]:
+        """
+        Returns `includes spark.databricks.clusterUsageTags.clusterAllTags` tags as `Dict`.
+        """
+        if not os.path.isfile(DATABRICKS_METRICS_PROP_PATH):
+            # We want to retry in case the cluster is still initializing, and the file is not yet deployed.
             return None
         webui = self.get_webui_address()
         # The API used: https://spark.apache.org/docs/latest/monitoring.html#rest-api
         apps_url = SPARKUI_APPS_URL.format(webui)
-        self.logger.debug(f"Databricks SparkUI address: {apps_url}.")
-        resp = requests.get(apps_url, timeout=REQUEST_TIMEOUT)
-        if not resp.ok:
-            self.logger.warning(
-                f"Failed initializing Databricks client. {apps_url!r} request failed, status_code: {resp.status_code}."
-            )
+        self.logger.debug("Databricks SparkUI address", apps_url=apps_url)
+        try:
+            response = self._request_get(apps_url)
+        except requests.exceptions.RequestException:
+            # Request might fail in cases where the cluster is still initializing, retrying.
             return None
-        apps = resp.json()
+        try:
+            apps = response.json()
+        except Exception as e:
+            if "Spark is starting up. Please wait a while until it's ready" in response.text:
```

Comment:
Hmm........... relying on plaintext is bad.
Don't this response have an HTTP status code ?

Reply:
We could use a separate API for that. If the response you get here is "try again" in HTTP codes, it is indeed better to test against that code.

Reply:
Bastards actually return 200OK lol : https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/SparkUI.scala#L64

Anyway it seems that after the service has started, the `initHandler` is removed and the `ApiRootHandler` (the json based) is added

Reply:
@roi-granulate good thing to add a reference to this source. With an absolute link not "master"

Reply:
64934a3